Fixes #25: Deletes non-discovered nfd labels

IRCody opened the pull request:
Took a stab at #25. Appreciate any feedback.

Modifies AddLabels() to first remove any non-discovered nfd labels from
the node labels before adding newly discovered labels.

Adds tests to main_test.go to test the behavior of AddLabels.

edit:

This deletes non-discovered labels including those that may not be "discovered" because of the whitelist option. Wanted to make sure I called that out.

ConnorDoyle commented on main.go:
```diff
@@ -232,6 +232,15 @@ func (h k8sHelpers) GetNode(cli *client.Client) (*api.Node, error) {
 }
 
 func (h k8sHelpers) AddLabels(n *api.Node, labels Labels) {
+	// Remove nfd node labels that no longer apply
```
Could you move this stuff out to another function? Removing labels in a function called `AddLabels` is just asking for trouble :)

IRCody replied:
Would it make sense to rename the function to UpdateLabels?

balajismaniam replied:
I would prefer changing `AddLabels` -> `AddorUpdateLabels` and moving the remove labels part to another function as Connor mentioned.

IRCody replied:
If you move the part that removes labels I don't think you would want to rename `AddLabels` as it's behavior hasn't changed.

Does it make sense to split this into another function? We'd be changing the set of calls to update labels to:
```go
helper.RemoveNFDLabels(node)
helper.AddLabels(node, labels)
```
instead of just doing:
```go
helper.UpdateNodeLabels(node, labels)
```

You can definitely do it that way but I'm not sure what value you gain by splitting it up since the behavior will always be done together.

balajismaniam replied:
Yes. I think it is better if we split it into another function. I would prefer calling it `RemoveLabels` similar to `AddLabels`.

balajismaniam replied:
Nit on the comment: We don't use `nfd` anywhere in the repo right now. Consider deleting `nfd` from the comment.

ConnorDoyle commented:
Thanks for taking this on @IRCody! @balajismaniam I unassigned myself since I'll be out. Would you mind reviewing & shepherding it?

balajismaniam commented:
@ConnorDoyle: Sure I can do that.

ConnorDoyle commented:
Testability seems better if they are split, so it gets my vote.

> On Oct 31, 2016, at 16:49, Cody Roseborough <notifications@github.com> wrote:
> 
> @IRCody commented on this pull request.
> 
> In main.go:
> 
> > @@ -232,6 +232,15 @@ func (h k8sHelpers) GetNode(cli *client.Client) (*api.Node, error) {
>  }
>  
>  func (h k8sHelpers) AddLabels(n *api.Node, labels Labels) {
> +	// Remove nfd node labels that no longer apply
> If you move the part that removes labels I don't think you would want to rename AddLabels as it's behavior hasn't changed.
> 
> Does it make sense to split this into another function? We'd be changing the set of calls to update labels to:
> 
> helper.RemoveNFDLabels(node)
> helper.AddLabels(node, labels)
> instead of just doing:
> 
> helper.UpdateNodeLabels(node, labels)
> You can definitely do it that way but I'm not sure what value you gain by splitting it up since the behavior will always be done together.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

balajismaniam commented on main.go:
```diff
@@ -232,6 +232,15 @@ func (h k8sHelpers) GetNode(cli *client.Client) (*api.Node, error) {
+	for k := range n.Labels {
+		// If the label is not in the labels we are adding
+		// and it is an nfd label, delete it.
+		if _, ok := labels[k]; !ok && strings.Contains(k, Namespace) {
```
s/Namespace/prefix/. See https://github.com/kubernetes-incubator/node-feature-discovery/pull/28#discussion_r85028520 for context.

balajismaniam commented on main_test.go:
```diff
@@ -100,3 +100,47 @@ func TestDiscoveryWithMockSources(t *testing.T) {
+		})
+
+		Convey("They should be added to the node.Labels", func() {
+			test1 := Namespace + ".test1"
```
Change `test1 := Namespace + ".test1"` to `test1 := prefix + "-test1"`.

balajismaniam commented on main_test.go:
```diff
@@ -100,3 +100,47 @@ func TestDiscoveryWithMockSources(t *testing.T) {
 }
+
+func TestAddLabels(t *testing.T) {
+
```
Nit: Delete this line.

balajismaniam commented on main_test.go:
```diff
@@ -100,3 +100,47 @@ func TestDiscoveryWithMockSources(t *testing.T) {
 	})
 }
+
+func TestAddLabels(t *testing.T) {
```
This should change appropriately as we are going to split remove and add into two functions.

balajismaniam commented on main_test.go:
```diff
@@ -100,3 +100,47 @@ func TestDiscoveryWithMockSources(t *testing.T) {
+			})
+
+			Convey("And they should be removed if not present on subsequent calls", func() {
+				test2 := Namespace + ".test2"
```
Change `test1 := Namespace + ".test2"` to `test1 := prefix + "-test2"`.

balajismaniam commented on main_test.go:
```diff
@@ -100,3 +100,47 @@ func TestDiscoveryWithMockSources(t *testing.T) {
+			So(n.Labels, ShouldContainKey, test1)
+
+			Convey("And they should be overwritten by subsequent calls", func() {
+				labels[test1] = "false"
```
Right now we don't support key values other than `true` or updates to already exisiting labels (other than removal after this PR gets merged).  Can we delete this test case?

balajismaniam commented on main.go:
```diff
@@ -232,6 +232,15 @@ func (h k8sHelpers) GetNode(cli *client.Client) (*api.Node, error) {
+	// Remove nfd node labels that no longer apply
+	for k := range n.Labels {
+		// If the label is not in the labels we are adding
+		// and it is an nfd label, delete it.
```
Nit: Consider changing comment to `If the label was not discovered in the current run of node-feature-discovery and if it has the prefix, delete it from the list of node labels` or something similar.

balajismaniam commented:
@IRCody As per our offline discussion, this PR will remove the node labels if the `--label-whitelist` changes between subsequent runs of the PR. I think we should update the readme to reflect this. We can do this in a separate PR if you prefer.

balajismaniam commented on main.go:
```diff
@@ -232,6 +232,15 @@ func (h k8sHelpers) GetNode(cli *client.Client) (*api.Node, error) {
+		// If the label is not in the labels we are adding
+		// and it is an nfd label, delete it.
+		if _, ok := labels[k]; !ok && strings.Contains(k, Namespace) {
+			delete(n.Labels, k)
```
Can we log the labels we are deleting here?

IRCody commented:
@balajismaniam: Updated this addressing the comments. A summary of what's changed:

* Split the Removal of labels into it's own function, `RemoveLabels` that takes in a the api.Node and a search string and removes all keys that contain the search string. This is called from the advertiseFeatureLabels() func with the `prefix` as you pointed out in your comments.
* Modified the test for AddLabels and added some for RemoveLabels
* Added a note in the README about the behavior around label removal and it's interaction with whitelist (or potentially any other label-modifying action that is added in the future).

balajismaniam commented on main_test.go:
```diff
@@ -100,3 +102,66 @@ func TestDiscoveryWithMockSources(t *testing.T) {
+			},
+		}
+
+		Convey("a unique label should be remove", func() {
```
s/remove/removed

balajismaniam commented on README.md:
```diff
@@ -80,6 +80,11 @@ the only label value published for features is the string `"true"`._
 The `--sources` flag controls which sources to use for discovery.
 
+_Note: consecutive runs of node-feature-discovery will update the labels on a
+given node. If labels are not discovered on a consecutive run they will be removed
```
s/If labels are not discovered on a consecutive run they will be removed/If features are not discovered on a consecutive run, the corresponding label will be removed

balajismaniam commented on README.md:
```diff
@@ -80,6 +80,11 @@ the only label value published for features is the string `"true"`._
+_Note: consecutive runs of node-feature-discovery will update the labels on a
+given node. If labels are not discovered on a consecutive run they will be removed
+from the node. This includes any restrictions placed on the consecutive run, such
+as restricting discovered nodes with the --label-whitelist option._
```
s/discovered nodes/discovered features/

balajismaniam commented:
@IRCody Left some minor comments. Otherwise, LGTM.

IRCody commented:
@balajismaniam: Made those updates.

balajismaniam commented:
@IRCody Thanks for the updates.

balajismaniam commented:
/lgtm
